dali/pipeline/data/buffer.h

```diff
@@ -307,7 +326,45 @@ class DLL_PUBLIC Buffer {
     return shares_data_;
   }
 
-  DISABLE_COPY_MOVE_ASSIGN(Buffer);
+  /**
+   * @brief Set another Buffer as the backing memory for this Buffer.
+   *
+   * Current Buffer will be marked as sharing data, and reallocation of memory will be
+   * prohibited until reset() is called.
+   */
+  inline void set_backing_allocation(const Buffer<Backend> &other) {
+    type_ = other.type_;
+    data_ = other.data_;
+    allocate_ = {};
+    size_ = other.size_;
+    shares_data_ = true;
+    num_bytes_ = other.num_bytes_;
+    device_ = other.device_;
+  }
+
+  /**
+   * @brief Set external memory as the backing memory for this Buffer.
+   *
+   * Current Buffer will be marked as sharing data, and reallocation of memory will be
+   * prohibited until reset() is called.
+   *
+   * For GPU memory, it is assumed to be associated with current device.
+   */
+  inline void set_backing_allocation(const shared_ptr<void> &ptr, size_t bytes,
+                                     DALIDataType type = DALI_NO_TYPE, size_t size = 0) {
+    type_ = TypeTable::GetTypeInfo(type);
+    data_ = ptr;
+    allocate_ = {};
+    size_ = size;
+    shares_data_ = true;
+    num_bytes_ = bytes;
+    // setting the allocation, get the device
+    if (std::is_same<Backend, GPUBackend>::value) {
+      CUDA_CALL(cudaGetDevice(&device_));
```

Comment:
I don't like this logic. Assuming, that this buffer already has a device - why would we change it? We have multiple sources of device_id - the one currently used, current device (as obtained from cudaGetDevice) and the real one - associated with the pointer. Unfortunately, getting the real one is too expensive - and it can be ambiguous (managed buffers, cuMemMap, etc). Since we can't get the real one, who's to say that current context holds the right device and not the buffer object?

Reply:
Maybe it should be somehow passed with the pointer? But if you are setting a totally new allocation (that is potentially a "3rd-party" allocation as you are just passing a pointer), I would assume that it's unrelated to what the Buffer currently holds. 
Either way let's try to limit ways of interleaving the setup of our memory objects, so we don't get another type-shape-reserve situation with all possible permutations.

Reply:
Adjusted by doing only when necessary and we have the CPU_ONLY device set.